Flaky test ldap.TestConnectWithTLS11

### Flaky Test: `ldap.TestConnectWithTLS11`

**Observed Failure Log (example):**
```
=== RUN   TestConnectWithTLS11
    ldap_common_test.go:141:
        Error Trace:  pkg/privilege/privileges/ldap/ldap_common_test.go:141
                        src/runtime/asm_amd64.s:1700
        Error:         Received unexpected error:
                       listen tcp 127.0.0.1:17302: bind: address already in use
        Test:          TestConnectWithTLS11
    ldap_common_test.go:182:
        Error Trace:  pkg/privilege/privileges/ldap/ldap_common_test.go:182
        Error:         Error "create ldap connection: LDAP Result Code 200 \"Network Error\": dial tcp [::1]:17302: connect: connection refused" does not contain "protocol version not supported"
        Test:          TestConnectWithTLS11
--- FAIL: TestConnectWithTLS11 (1.99s)
```

**Symptoms:**
1. Listener fails to bind with `address already in use`.
2. Test then continues and expects error containing `"protocol version not supported"` but instead gets a connection-refused wrapped in the LDAP network error.
3. Subsequent assertions fail; in some runs this can cascade into a panic due to assumptions about initialized resources.

**Root Cause:**
The test chooses a pseudo-random port using:
```go
randomTLSServicePort := rand.Int()%10000 + 10000
ln, err = tls.Listen("tcp", fmt.Sprintf("localhost:%d", randomTLSServicePort), tlsConfig)
```
This is racy/flaky because other concurrent tests or processes may already occupy the chosen port between the time it is selected and `Listen` executes. When the bind fails, the goroutine closes the channel signaling readiness, the main test path proceeds as if the server were available, and the client dial fails with connection refused.

**Why It’s Flaky Now:**
Parallel / randomized test execution or system background services occasionally collide with the selected port range (10000–19999). The timing window is small but non-zero, hence intermittent failures.

**Proposed Fix:**
Adopt the same pattern already used in `TestConnectThrough636`: let the OS allocate a free ephemeral port by listening on `"localhost:0"`, then extract the actual port from `ln.Addr()`. This guarantees a free port at bind time.

**Suggested Patch (conceptual diff):**
```diff
 func TestConnectWithTLS11(t *testing.T) {
-    // this test only tests whether the LDAP with LTS enabled will fallback from StartTLS
-    randomTLSServicePort := rand.Int()%10000 + 10000
+    // Use an ephemeral port to avoid flakiness from manual random selection.
+    // This mirrors the strategy in TestConnectThrough636.
     serverWg := &sync.WaitGroup{}
     serverWg.Add(1)
     go func() {
         defer close(startListen)
         defer serverWg.Done()
         cert, err := tls.X509KeyPair(tlsCrtStr, tlsKeyStr)
         require.NoError(t, err)
         tlsConfig := &tls.Config{ Certificates: []tls.Certificate{cert}, MaxVersion: tls.VersionTLS11 }
-        ln, err = tls.Listen("tcp", fmt.Sprintf("localhost:%d", randomTLSServicePort), tlsConfig)
+        ln, err = tls.Listen("tcp", "localhost:0", tlsConfig)
         require.NoError(t, err)
         startListen <- struct{}{}
         ...
     }()
 
     <-startListen
     defer func() { ... }()
 
     impl := &ldapAuthImpl{}
     impl.SetEnableTLS(true)
     impl.SetLDAPServerHost("localhost")
-    impl.SetLDAPServerPort(randomTLSServicePort)
+    // Derive the chosen port actually in use
+    _, portStr, err := net.SplitHostPort(ln.Addr().String())
+    require.NoError(t, err)
+    p, err := strconv.Atoi(portStr)
+    require.NoError(t, err)
+    impl.SetLDAPServerPort(p)
```

**Optional Additional Improvement:**
Add a guard that if `ln` is nil after startup signaling, fail early with a clearer message (though using `:0` makes this unlikely to trigger).

**Action Items:**
1. Implement the patch above.
2. Remove the unused `rand` import from the test file if it becomes unused.
3. Run the LDAP-related test suite multiple times (e.g. `go test -count=50 ./pkg/privilege/privileges/ldap -run TestConnectWithTLS11`) to ensure flakiness is resolved.

**Impact:**
Eliminates intermittent CI failures and reduces noise in test results.

Assigning to @bb7133 for confirmation and patch submission.

/cc @bb7133

--- a/pkg/privilege/privileges/ldap/ldap_common_test.go
+++ b/pkg/privilege/privileges/ldap/ldap_common_test.go
@@ -19,8 +19,6 @@ import (
 	"crypto/tls"
 	"crypto/x509"
 	_ "embed"
-	"fmt"
-	"math/rand"
 	"net"
 	"strconv"
 	"sync"
@@ -121,25 +119,26 @@ func TestConnectThrough636(t *testing.T) {
 func TestConnectWithTLS11(t *testing.T) {
 	var ln net.Listener
 
-	startListen := make(chan struct{})
-
-	// this test only tests whether the LDAP with LTS enabled will fallback from StartTLS
-	randomTLSServicePort := rand.Int()%10000 + 10000
+	startListen := make(chan error, 1)
 	serverWg := &sync.WaitGroup{}
 	serverWg.Add(1)
 	go func() {
-		defer close(startListen)
 		defer serverWg.Done()
 
 		cert, err := tls.X509KeyPair(tlsCrtStr, tlsKeyStr)
-		require.NoError(t, err)
+		if err != nil {
+			startListen <- err
+			return
+		}
 		tlsConfig := &tls.Config{
 			Certificates: []tls.Certificate{cert},
 			MaxVersion:   tls.VersionTLS11,
 		}
-		ln, err = tls.Listen("tcp", fmt.Sprintf("localhost:%d", randomTLSServicePort), tlsConfig)
-		require.NoError(t, err)
-		startListen <- struct{}{}
+		ln, err = tls.Listen("tcp", "localhost:0", tlsConfig)
+		startListen <- err
+		if err != nil {
+			return
+		}
 
 		for {
 			conn, err := ln.Accept()
@@ -164,7 +163,7 @@ func TestConnectWithTLS11(t *testing.T) {
 		}
 	}()
 
-	<-startListen
+	require.NoError(t, <-startListen)
 	defer func() {
 		require.NoError(t, ln.Close())
 		serverWg.Wait()
@@ -173,12 +172,16 @@ func TestConnectWithTLS11(t *testing.T) {
 	impl := &ldapAuthImpl{}
 	impl.SetEnableTLS(true)
 	impl.SetLDAPServerHost("localhost")
-	impl.SetLDAPServerPort(randomTLSServicePort)
+	_, port, err := net.SplitHostPort(ln.Addr().String())
+	require.NoError(t, err)
+	p, err := strconv.Atoi(port)
+	require.NoError(t, err)
+	impl.SetLDAPServerPort(p)
 
 	impl.caPool = x509.NewCertPool()
 	require.True(t, impl.caPool.AppendCertsFromPEM(tlsCAStr))
 
-	_, err := impl.connectionFactory()
+	_, err = impl.connectionFactory()
 	require.ErrorContains(t, err, "protocol version not supported")
 }
 
@@ -197,16 +200,18 @@ func TestLDAPStartTLSTimeout(t *testing.T) {
 	defer close(afterTimeout)
 
 	// this test only tests whether the LDAP with LTS enabled will fallback from StartTLS
-	randomTLSServicePort := rand.Int()%10000 + 10000
+	startListenErr := make(chan error, 1)
 	serverWg := &sync.WaitGroup{}
 	serverWg.Add(1)
 	go func() {
 		var err error
-		defer close(startListen)
 		defer serverWg.Done()
 
-		ln, err = net.Listen("tcp", fmt.Sprintf("localhost:%d", randomTLSServicePort))
-		require.NoError(t, err)
+		ln, err = net.Listen("tcp", "localhost:0")
+		startListenErr <- err
+		if err != nil {
+			return
+		}
 		startListen <- struct{}{}
 
 		conn, err := ln.Accept()
@@ -219,6 +224,7 @@ func TestLDAPStartTLSTimeout(t *testing.T) {
 		require.NoError(t, ln.Close())
 	}()
 
+	require.NoError(t, <-startListenErr)
 	<-startListen
 	defer func() {
 		serverWg.Wait()
@@ -227,7 +233,11 @@ func TestLDAPStartTLSTimeout(t *testing.T) {
 	impl := &ldapAuthImpl{}
 	impl.SetEnableTLS(true)
 	impl.SetLDAPServerHost("localhost")
-	impl.SetLDAPServerPort(randomTLSServicePort)
+	_, port, err := net.SplitHostPort(ln.Addr().String())
+	require.NoError(t, err)
+	p, err := strconv.Atoi(port)
+	require.NoError(t, err)
+	impl.SetLDAPServerPort(p)
 
 	impl.caPool = x509.NewCertPool()
 	require.True(t, impl.caPool.AppendCertsFromPEM(tlsCAStr))
@@ -235,7 +245,7 @@ func TestLDAPStartTLSTimeout(t *testing.T) {
 	impl.SetMaxCapacity(1)
 
 	now := time.Now()
-	_, err := impl.connectionFactory()
+	_, err = impl.connectionFactory()
 	afterTimeout <- struct{}{}
 	dur := time.Since(now)
 	require.Greater(t, dur, 2*time.Second)

